packages/databricks-vscode/src/bundle/BundleRunManager.ts

```diff
@@ -0,0 +1,184 @@
+                this.createTerminal(terminalName)
+            );
+        }
+        let terminal = this.terminalDetails.get(terminalName)!;
+
+        const disposables: Disposable[] = [];
+        try {
+            terminal.terminal.show();
+            if (
+                window.terminals.find(
+                    (i) => i.name === terminal?.terminal.name
+                ) === undefined
+            ) {
+                // The terminal has been closed. Recreate everything.
+                terminal = this.createTerminal(terminalName);
+                this.terminalDetails.set(terminalName, terminal);
+            }
+            if (terminal.pty.process !== undefined) {
+                // There is already a process running. Raise error
+                throw new Error(
+                    `Process already running. Pid: ${terminal.pty.process.pid}`
+                );
+            }
+            this.cancellationTokenSources.get(terminalName)?.cancel();
+            this.cancellationTokenSources.get(terminalName)?.dispose();
+            this.cancellationTokenSources.delete(terminalName);
+
+            const cancellationTokenSource = new CancellationTokenSource();
+            this.cancellationTokenSources.set(
+                terminalName,
+                cancellationTokenSource
+            );
+            const onCancellationEvent =
+                cancellationTokenSource.token.onCancellationRequested(() => {
+                    terminal?.pty.close();
+                    //Dispose self on cancellation
+                    onCancellationEvent.dispose();
+                });
+
+            const cmd = this.bundleRemoteStateModel.getRunCommand(resourceKey);
+
+            await this.isRunningMutex.synchronise(async () => {
+                this._isRunning.set(terminalName, true);
+                this.onDidChangeEmitter.fire();
+            });
+            // spawn a new process with the latest command, in the same terminal.
+            terminal.pty.spawn(cmd);
+            terminal.terminal.show();
+
+            // Wait for the process to exit
+            await new Promise<void>((resolve, reject) => {
+                if (terminal === undefined) {
+                    resolve();
+                    return;
+                }
+                disposables.push(
+                    terminal.pty.onDidCloseProcess(
+                        (exitCode) => {
+                            if (exitCode === 0) {
+                                resolve();
+                            } else {
+                                reject(
+                                    new Error(
+                                        `Process exited with code ${exitCode}`
+                                    )
+                                );
+                            }
+                        },
+                        window.onDidCloseTerminal((e) => {
+                            e.name === terminal.terminal.name &&
```

Comment:
Should we just compare refs? E.g. `e === terminal.terminal`? Shouldn't change much, but since we are re-using terminal names we might end up with some bugs were it will be relevant

Reply:
We want to have the same terminal name as well. I was keying terminal names with it. I think there was some issue with using references. They are not the same somehow.